firedrake/preconditioners/base.py

```diff
@@ -109,25 +109,21 @@ def get_appctx(pc):
         return get_appctx(pc.getDM()).appctx
 
     @staticmethod
-    def new_snes_ctx(pc, op, bcs, mat_type, fcp=None, options_prefix=None):
+    def new_snes_ctx(pc, op, bcs, mat_type, fcp=None, options_prefix=None, pre_apply_bcs=True):
         """ Create a new SNES contex for nested preconditioning
         """
```

Comment:
```suggestion
        """Create a new `_SNESContext` for nested (linear) preconditioning
        
        Parameters
        ----------
        pc : PETSc.PC
             The PC object.
        op : ufl.BaseForm
             A bilinear form.
        bcs : DirichletBC[]
             The boundary conditions.
        mat_type : str
             The matrix type for the assembly of ``op``.
        options_prefix : str
             The PETSc options prefix for the new `_SNESContext`.
        pre_apply_bcs : bool
             If True, the ``bcs`` are pre applied on the solution before the solve, 
             otherwise the residual of the ``bcs`` is included in the linear system.
        
        Returns
        -------
        A new `_SNESContext`
        """
```

Reply:
@JHopeCollins feel free to add more documentation

Reply:
Great, much better. Thanks @pbrubeck!